src/crystal/system/unix/file.cr

```diff
@@ -70,28 +70,28 @@ module Crystal::System::File
     tmpdir.rchop(::File::SEPARATOR)
   end
 
-  def self.stat(path)
+  def self.stat?(path : String) : ::File::Stat?
     if LibC.stat(path.check_no_null_byte, out stat) != 0
-      raise Errno.new("Unable to get stat for '#{path}'")
+      if {Errno::ENOENT, Errno::ENOTDIR}.includes? Errno.value
```

Comment:
Previously, `File.file?("/path/to/file.cr/some/rubbish")` would raise `ENOTDIR` because `/path/to/file.cr` exists. I've changed that in this PR to allow `Dir.exists?` to be implemented in terms of `File.stat?`. Dir does not raise on the above condition, which is specified [in the specs](https://github.com/crystal-lang/crystal/blob/master/spec/std/dir_spec.cr#L30). `File.file?` does.

Ruby implements it how i've implemented it here:
```
$ ruby -e 'p File.file?("/etc/fstab/foo")'
false
$ crystal eval 'p File.file?("/etc/fstab/foo")'
stat: Not a directory (Errno)
$ bin/crystal eval 'p File.file?("/etc/fstab/foo")'
false
```